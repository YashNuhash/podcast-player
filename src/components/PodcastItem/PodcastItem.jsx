import React from 'react'
import { Link } from 'react-router-dom'
import EllipsisText from '../EllipsisText/EllipsisText'
import styles from './PodcastItem.module.scss'

export const PodcastItem = ({ id, thumbnail, title, publisher }) => (
  <div className={styles.Item} key={id}>
    <div className={styles.ImageWrapper}>
      <img className={styles.Image} src={thumbnail} alt='' />
    </div>
    <div className={styles.Content}>
      <EllipsisText tagName='p' className={styles.Publisher}>
        {publisher}
      </EllipsisText>
      <EllipsisText tagName='h4' className={styles.Title}>
        {title}
      </EllipsisText>

      <Link to={`/podcast/${id}`} className={styles.Button}>
        View More
      </Link>
    </div>
  </div>
)

export default PodcastItem
