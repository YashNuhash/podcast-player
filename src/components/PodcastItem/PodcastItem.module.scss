@import '../../utils';

.Item {
    display: flex;
}

.Content {
    display: flex;
    flex-direction: column;
    justify-content: space-evenly;
}

.ImageWrapper {
    height: 15rem;
    width: 15rem;
    background-color: $bg;
    border-radius: 3rem;
    overflow: hidden;
    margin-right: 3rem;
    position: relative;
    flex-shrink: 0;
}

.Image {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
}

.Title,
.Publisher {
    display: -webkit-box;
    -webkit-box-orient: vertical;
    -webkit-line-clamp: 2;
    text-overflow: ellipsis;
    white-space: normal;
    overflow: hidden;
}

.Publisher {
    color: $light-text;
    font-weight: 600;
    font-size: 1.4rem;
}

.Button {
    padding: 0.5rem 1rem;
    text-align: center;
    border: 1px solid $primary;
    border-radius: 100px;
    color: $primary;
    text-decoration: none;
    align-self: flex-start;
    font-size: 1.5rem;
}

.Slider {
    flex-direction: column;

    .Image {
        margin-right: 0;
        margin-bottom: 1rem;
    }

    .Publisher {
        font-size: 1.3rem;
    }

    .Title {
        font-size: 1.5rem;
    }
}
