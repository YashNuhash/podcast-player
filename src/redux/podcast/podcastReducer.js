import PodcastTypes from './podcastTypes'

const initialState = {
    podcast: null,
    recommendations: null,
    isPodcastFetching: false,
    areRecommendationsFetching: false,
    areEpisodesFetching: false,
    error: null,
}

const podcastReducer = (state = initialState, action) => {
    switch (action.type) {
        case PodcastTypes.FETCH_PODCAST_START:
            return {
                ...state,
                isPodcastFetching: true,
            }
        case PodcastTypes.FETCH_PODCAST_SUCCESS:
            return {
                ...state,
                isPodcastFetching: false,
                podcast: action.payload,
                error: null,
            }

        case PodcastTypes.FETCH_RECOMMENDATIONS_START:
            return { ...state, areRecommendationsFetching: true }
        case PodcastTypes.FETCH_RECOMMENDATIONS_SUCCESS:
            return {
                ...state,
                areRecommendationsFetching: false,
                recommendations: action.payload,
            }

        case PodcastTypes.FETCH_MORE_EPISODES_START:
            return { ...state, areEpisodesFetching: true }
        case PodcastTypes.FETCH_MORE_EPISODES_SUCCESS:
            return {
                ...state,
                areEpisodesFetching: false,
                podcast: {
                    ...action.payload,
                    episodes: [
                        ...state.podcast.episodes,
                        ...action.payload.episodes,
                    ],
                },
            }

        case PodcastTypes.FETCH_PODCAST_FAILURE:
        case PodcastTypes.FETCH_RECOMMENDATIONS_FAILURE:
        case PodcastTypes.FETCH_MORE_EPISODES_FAILURE:
            return {
                ...state,
                error: action.payload,
            }
        default:
            return state
    }
}

export default podcastReducer
