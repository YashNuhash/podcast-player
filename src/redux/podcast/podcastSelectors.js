import { createSelector } from 'reselect'

export const podcast = state => state.podcast

export const selectPodcast = createSelector(
    [podcast],
    podcast => podcast.podcast,
)

export const selectRecommendations = createSelector(
    [podcast],
    podcast => podcast.recommendations,
)

export const selectIsPodcastFetching = createSelector(
    [podcast],
    podcast => podcast.isPodcastFetching,
)

export const selectAreRecommendationsFetching = createSelector(
    [podcast],
    podcast => podcast.areRecommendationsFetching,
)

export const selectAreEpisodesFetching = createSelector(
    [podcast],
    podcast => podcast.areEpisodesFetching,
)

export const selectEpisodes = createSelector(
    [podcast],
    podcast => podcast.podcast.episodes,
)

export const selectAreMoreEpisodes = createSelector(
    [podcast],
    podcast => !!podcast.podcast.next_episode_pub_date,
)
