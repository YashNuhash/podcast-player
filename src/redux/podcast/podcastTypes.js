export const START_FETCH_PODCAST = 'podcast/START_FETCH_PODCAST'
export const SET_PODCAST = 'podcast/SET_PODCAST'

export const START_FETCH_MORE_EPISODES = 'podcast/START_FETCH_MORE_EPISODES'
export const SET_MORE_EPISODES = 'podcast/SET_MORE_EPISODES'

export const START_FETCH_RECOMMENDATIONS = 'podcast/START_FETCH_RECOMMENDATIONS'
export const SET_RECOMMENDATIONS = 'podcast/SET_RECOMMENDATIONS'
