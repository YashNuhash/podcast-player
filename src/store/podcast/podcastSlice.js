import { createSlice } from '@reduxjs/toolkit'
import { listenNotesApi } from '../../axios'

export const initialState = {
  podcast: null,
  recommendations: null,
  isPodcastFetching: false,
  areRecommendationsFetching: false,
  areEpisodesFetching: false,
  error: null,
}

export const podcastSlice = createSlice({
  name: 'podcast',
  initialState,
  reducers: {
    setLoading(state) {
      state.isPodcastFetching = true
    },
    setPodcast(state, action) {
      state.isPodcastFetching = false
      state.podcast = action.payload
      state.error = null
    },
    setError(state, action) {
      state.error = action.payload
    },
    setRecommendationsLoading(state) {
      state.areRecommendationsFetching = true
    },
    setRecommendations(state, action) {
      state.areRecommendationsFetching = false
      state.recommendations = action.payload
      state.error = null
    },
    setEpisodesLoading(state) {
      state.areEpisodesFetching = true
    },
    setEpisodes(state, action) {
      const episodes = state.podcast.episodes
      state.areEpisodesFetching = false
      state.podcast = action.payload
      state.podcast.episodes = [...episodes, ...action.payload.episodes]
      state.error = null
    },
  },
})

export const {
  setLoading,
  setPodcast,
  setError,
  setRecommendationsLoading,
  setRecommendations,
  setEpisodesLoading,
  setEpisodes,
} = podcastSlice.actions

export const fetchPodcast = async (dispatch, podcastId) => {
  try {
    dispatch(setLoading())

    const { data } = await listenNotesApi.get(`/podcasts/${podcastId}`)

    dispatch(setPodcast(data))
  } catch (error) {
    dispatch(setError(error))
  }
}

export const fetchRecommendations = async (dispatch, podcastId) => {
  try {
    dispatch(setRecommendationsLoading())

    const { data } = await listenNotesApi.get(`/podcasts/${podcastId}/recommendations`)
    dispatch(setRecommendations(data.recommendations))
  } catch (error) {
    dispatch(setError(error))
  }
}

export const fetchEpisodes = async (dispatch, podcastId, next_episode_pub_date) => {
  try {
    dispatch(setEpisodesLoading())

    const { data } = await listenNotesApi.get(
      `/podcasts/${podcastId}?next_episode_pub_date=${next_episode_pub_date}`,
    )
    dispatch(setEpisodes(data))
  } catch (error) {
    dispatch(setError(error))
  }
}

export default podcastSlice.reducer
