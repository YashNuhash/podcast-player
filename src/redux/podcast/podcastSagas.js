import { all, call, takeLatest, put, select } from 'redux-saga/effects'
import PodcastTypes from './podcastTypes'
import {
    fetchPodcastStart,
    fetchPodcastSuccess,
    fetchPodcastFailure,
    fetchRecommendationsStart,
    fetchRecommendationsSuccess,
    fetchRecommendationsFailure,
    fetchMoreEpisodesStart,
    fetchMoreEpisodesSuccess,
    fetchMoreEpisodesFailure,
} from './podcastActions'
import { listenNotesApi } from '../../axios'
import { podcast } from './podcastSelectors'

export function* fetchPodcastAsync({ payload: podcastId }) {
    try {
        yield put(fetchPodcastStart())

        const { data } = yield listenNotesApi.get(`/podcasts/${podcastId}`)

        yield put(fetchPodcastSuccess(data))
    } catch (error) {
        yield put(fetchPodcastFailure(error))
    }
}

export function* fetchRecommendationsAsync({ payload: podcastId }) {
    try {
        yield put(fetchRecommendationsStart())

        const { data } = yield listenNotesApi.get(
            `/podcasts/${podcastId}/recommendations`,
        )

        yield put(fetchRecommendationsSuccess(data.recommendations))
    } catch (error) {
        yield put(fetchRecommendationsFailure(error))
    }
}

export function* fetchMoreEpisodesAsync() {
    try {
        yield put(fetchMoreEpisodesStart())

        const {
            podcast: { id, next_episode_pub_date },
        } = yield select(podcast)

        const { data } = yield listenNotesApi.get(
            `/podcasts/${id}?next_episode_pub_date=${next_episode_pub_date}`,
        )
        yield put(fetchMoreEpisodesSuccess(data))
    } catch (error) {
        yield put(fetchMoreEpisodesFailure(error))
    }
}

export function* onFetchPodcast() {
    yield takeLatest(PodcastTypes.FETCH_PODCAST, fetchPodcastAsync)
}

export function* onFetchRecommendations() {
    yield takeLatest(
        PodcastTypes.FETCH_RECOMMENDATIONS,
        fetchRecommendationsAsync,
    )
}

export function* onFetchMoreEpisodes() {
    yield takeLatest(PodcastTypes.FETCH_MORE_EPISODES, fetchMoreEpisodesAsync)
}

export function* podcastSagas() {
    yield all([
        call(onFetchPodcast),
        call(onFetchRecommendations),
        call(onFetchMoreEpisodes),
    ])
}
