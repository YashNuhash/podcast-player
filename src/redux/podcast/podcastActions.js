import { PLAY_EPISODE, TOGGLE_PLAY, SET_VOLUME } from './podcastTypes'

export const playEpisode = (podcastId, episode) => ({
    type: PLAY_EPISODE,
    payload: {
        podcastId,
        episode,
    },
})

export const setVolume = volume => ({
    type: SET_VOLUME,
    payload: volume,
})

export const togglePlay = () => ({
    type: TOGGLE_PLAY,
})
