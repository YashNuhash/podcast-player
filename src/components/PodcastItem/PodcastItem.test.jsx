import React from 'react'
import { MemoryRouter } from 'react-router-dom'
import { render, cleanup } from '@testing-library/react'
import { PodcastItem } from './PodcastItem'
import podcast from '../../fixtures/podcast'

afterEach(cleanup)

describe('PodcastItem', () => {
  test('match snapshot', () => {
    const { getByText } = render(
      <MemoryRouter>
        <PodcastItem {...podcast} />
      </MemoryRouter>,
    )
    getByText(podcast.title)
    getByText(podcast.publisher)
  })
})
