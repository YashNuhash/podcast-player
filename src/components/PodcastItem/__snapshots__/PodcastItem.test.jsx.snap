// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PodcastItem match snapshot 1`] = `
<DocumentFragment>
  <div
    class="Item"
  >
    <div
      class="ImageWrapper"
    >
      <img
        alt=""
        class="Image"
        src="https://cdn-images-1.listennotes.com/podcasts/unfictional-kcrw-SqjEeFXNKk7-RiDjOlNuLve.300x300.jpg"
      />
    </div>
    <div
      class="Content"
    >
      <p
        class="Ellipsis Publisher"
      >
        KCRW
      </p>
      <h4
        class="Ellipsis Title"
      >
        UnFictional
      </h4>
      <a
        class="Button"
        href="/podcast/a5f2e7e4897847139be47e84e181c674"
      >
        View More
      </a>
    </div>
  </div>
</DocumentFragment>
`;
